### PR TITLE
[Cart][Core] Rework the logic behind the cart

### PR DESCRIPTION
**Pull request by Arminek**
| Q               | A
| --------------- | ---
| Bug fix?        | yes
| New feature?    | yes
| BC breaks?      | yes
| Related tickets | https://github.com/Sylius/Sylius/pull/5197
| License         | MIT

### CODE REVIEW

File: src/Sylius/Bundle/CartBundle/EventListener/SessionCartSubscriber.php [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        $request = $event->getRequest();
+        // Hacky hack. Until there is a better solution.
+        if (!$this->isHtmlRequest($request)) {
+            return;
+        }
+
+        try {
+            $cart = $this->cartContext->getCart();
+        } catch (CartNotFoundException $exception) {
+            $cart = null;
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/CartBundle/EventListener/SessionCartSubscriber.php]: return?

**Comment by Mr-Negative** [comment]
To have it more visible. Any chances / maybe it is worked on #5566

**Comment by Arminek** [comment]
@Mr-Negative IMO we should rework basic cart logic I don't think that multiple orders are our goal now.

File: src/Sylius/Bundle/CoreBundle/Resources/config/context.xml [hunk 2]
```diff
@@ -57,4 +63,5 @@
             </call>
         </service>
     </services>
+
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/CoreBundle/Resources/config/context.xml]: Redundant

File: src/Sylius/Component/Core/Cart/Context/ShopBasedCartContext.php [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    {
+        if (null !== $this->cart) {
+            return $this->cart;
+        }
+
+        /** @var OrderInterface $cart */
+        $cart = $this->cartFactory->createNew();
+
+        if (null === $cart) {
+            throw new \LogicException('Decorated cart provider must return a cart instance, null given.');
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Core/Cart/Context/ShopBasedCartContext.php]: No need for that, factory will never return `null`

File: src/Sylius/Component/Core/Cart/Context/ShopBasedCartContext.php [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        }
+
+        try {
+            $cart->setChannel($this->shopperContext->getChannel());
+        } catch (ChannelNotFoundException $exception) {
+            throw new CartNotFoundException($exception);
+        }
+        $cart->setCustomer($this->shopperContext->getCustomer());
+        $cart->setCurrencyCode($this->shopperContext->getCurrencyCode());
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Core/Cart/Context/ShopBasedCartContext.php]: I would move this line into try-catch block since it also can throw `ChannelNotFoundException`.

File: src/Sylius/Component/Core/Cart/Context/ShopBasedCartContext.php [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        $cart = $this->cartFactory->createNew();
+
+        if (null === $cart) {
+            throw new \LogicException('Decorated cart provider must return a cart instance, null given.');
+        }
+
+        try {
+            $cart->setChannel($this->shopperContext->getChannel());
+        } catch (ChannelNotFoundException $exception) {
+            throw new CartNotFoundException($exception);
```
Comment (pamil) [inline src/Sylius/Component/Core/Cart/Context/ShopBasedCartContext.php]: :+1:

File: src/Sylius/Component/Core/Cart/Context/ShopBasedCartContext.php [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+        $this->shopperContext = $shopperContext;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getCart()
+    {
+        if (null !== $this->cart) {
+            return $this->cart;
```
Comment (pamil) [inline src/Sylius/Component/Core/Cart/Context/ShopBasedCartContext.php]: If we really DO need this "caching" feature, we should extract it as a decorator for `CartContextInterface`.

File: src/Sylius/Bundle/CartBundle/Context/SessionBasedCartContext.php [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+        $this->cartRepository = $cartRepository;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getCart()
+    {
+        if (!$this->session->has($this->sessionKeyName)) {
+            return null;
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/CartBundle/Context/SessionBasedCartContext.php]: According to the interface we should never return `null`, but throw exception.

File: src/Sylius/Bundle/CartBundle/Context/SessionBasedCartContext.php [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+        }
+
+        $cart = $this->cartRepository->findCartById($this->session->get($this->sessionKeyName));
+
+        if (null === $cart) {
+            $this->session->remove($this->sessionKeyName);
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/CartBundle/Context/SessionBasedCartContext.php]: Exception should be thrown after that.

File: src/Sylius/Component/Core/Cart/Context/ShopBasedCartContext.php (hunk 6 above)
Comment (pjedrzejewski) [inline src/Sylius/Component/Core/Cart/Context/ShopBasedCartContext.php]: @pamil This is not for caching. You want a single "new" cart per request, right? Because you call ``getCart()`` multiple times when adding for example and then saving. So it is not really for perf. reasons, but for the way that cart context is supposed to work.

File: src/Sylius/Bundle/CoreBundle/EventListener/CartBlamerListener.php [hunk 9]
```diff
@@ -90,16 +88,16 @@ private function blame(UserInterface $user)
...
-        $cart = $this->cartProvider->getCart();
-
         if (!$cart instanceof OrderInterface) {
             throw new UnexpectedTypeException($cart, OrderInterface::class);
         }
 
         return $cart;
     }
-}
+}
```
Comment (tuka217) [inline src/Sylius/Bundle/CoreBundle/EventListener/CartBlamerListener.php]: Missing blank line

File: src/Sylius/Component/Cart/spec/Context/CompositeCartContextSpec.php [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+
+namespace spec\Sylius\Component\Cart\Context;
+
+use PhpSpec\ObjectBehavior;
+use Sylius\Component\Cart\Context\CartContextInterface;
+use Sylius\Component\Cart\Context\CartNotFoundException;
+use Sylius\Component\Cart\Model\CartInterface;
+use Sylius\Component\Registry\PrioritizedServiceRegistryInterface;
+use Zend\Stdlib\PriorityQueue;
+use Zend\Stdlib\SplPriorityQueue;
```
Comment (NeverResponse) [inline src/Sylius/Component/Cart/spec/Context/CompositeCartContextSpec.php]: Unused

File: src/Sylius/Bundle/CartBundle/Context/SessionBasedCartContext.php [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+        $this->cartRepository = $cartRepository;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getCart()
+    {
+        if (!$this->session->has($this->sessionKeyName)) {
+            throw new CartNotFoundException();
```
Comment (pamil) [inline src/Sylius/Bundle/CartBundle/Context/SessionBasedCartContext.php]: Can we provide more information about what happened?

File: src/Sylius/Bundle/CartBundle/Context/SessionBasedCartContext.php [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+        }
+
+        $cart = $this->cartRepository->findCartById($this->session->get($this->sessionKeyName));
+
+        if (null === $cart) {
+            $this->session->remove($this->sessionKeyName);
+
+            throw new CartNotFoundException();
```
Comment (pamil) [inline src/Sylius/Bundle/CartBundle/Context/SessionBasedCartContext.php]: Can we provide more information about what happened?

File: src/Sylius/Bundle/CartBundle/Context/SessionBasedCartContext.php (hunk 8 above)
Comment (pamil) [inline src/Sylius/Bundle/CartBundle/Context/SessionBasedCartContext.php]: It is now :)

File: src/Sylius/Bundle/CartBundle/Controller/Controller.php [hunk 13]
```diff
@@ -58,19 +58,19 @@ protected function getCartSummaryRoute()
     protected function getCurrentCart()
     {
         return $this
-            ->getProvider()
+            ->getContext()
             ->getCart()
         ;
```
Comment (pamil) [inline src/Sylius/Bundle/CartBundle/Controller/Controller.php]: I guess we can make it an one-liner.

File: src/Sylius/Bundle/CartBundle/Controller/Controller.php [hunk 14]
```diff
@@ -58,19 +58,19 @@ protected function getCartSummaryRoute()
...
     }
 
     /**
-     * Get cart provider.
+     * Get cart context.
```
Comment (pamil) [inline src/Sylius/Bundle/CartBundle/Controller/Controller.php]: Quite redundant :tada:

File: src/Sylius/Bundle/CartBundle/DependencyInjection/Compiler/RegisterCartContextsPass.php [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+namespace Sylius\Bundle\CartBundle\DependencyInjection\Compiler;
+
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Reference;
+
+/**
+ * @author Arkadiusz Krakowiak <arkadiusz.krakowiak@lakion.com>
+ */
+class RegisterCartContextsPass implements CompilerPassInterface
```
Comment (pamil) [inline src/Sylius/Bundle/CartBundle/DependencyInjection/Compiler/RegisterCartContextsPass.php]: `final`?

File: src/Sylius/Bundle/CoreBundle/Resources/config/doctrine/model/Channel.orm.xml [hunk 16]
```diff
@@ -24,10 +24,10 @@
             <join-column name="default_locale_id" referenced-column-name="id" nullable="true" on-delete="SET NULL"/>
         </many-to-one>
         <many-to-one field="defaultCurrency" target-entity="Sylius\Component\Currency\Model\CurrencyInterface">
-            <join-column name="default_currency_id" referenced-column-name="id" nullable="false"/>
+            <join-column name="default_currency_id" referenced-column-name="id" nullable="true" on-delete="SET NULL" />
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/CoreBundle/Resources/config/doctrine/model/Channel.orm.xml]: Should be reverted.

File: src/Sylius/Bundle/CartBundle/EventListener/SessionCartSubscriber.php [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+     */
+    public function onKernelResponse(FilterResponseEvent $event)
+    {
+        if (!$event->isMasterRequest()) {
+            return;
+        }
+
+        /** @var Request $request */
+        $request = $event->getRequest();
+        // Hacky hack. Until there is a better solution.
```
Comment (pamil) [inline src/Sylius/Bundle/CartBundle/EventListener/SessionCartSubscriber.php]: What's the origin of this hack? What do we want to avoid?

File: src/Sylius/Bundle/CartBundle/Resources/config/services.xml [hunk 18]
```diff
@@ -17,33 +17,42 @@
...
         <parameter key="sylius.listener.cart_flash.class">Sylius\Bundle\CartBundle\EventListener\FlashSubscriber</parameter>
+        <parameter key="sylius.listener.session_cart.class">Sylius\Bundle\CartBundle\EventListener\SessionCartSubscriber</parameter>
 
         <parameter key="sylius.cart.purger.class">Sylius\Bundle\CartBundle\Purger\ExpiredCartsPurger</parameter>
     </parameters>
 
     <services>
-        <service id="sylius.context.cart" class="%sylius.context.cart.class%">
-            <argument /> <!-- Storage service name comes from config -->
+        <service id="sylius.registry.cart_context" class="Sylius\Component\Registry\PrioritizedServiceRegistry">
```
Comment (pamil) [inline src/Sylius/Bundle/CartBundle/Resources/config/services.xml]: Registry component it's quite weird, prioritized service registry is not a registry but a prioritized vector. I'd rather use `zend/stdlib` and `Zend\Stdlib\PriorityQueue` inside a composite context (look at `CompositeChannelContext`, it's quite simple). Moreover, we don't use it anywhere, but in composite context.

File: src/Sylius/Bundle/CartBundle/spec/EventListener/CartSubscriberSpec.php [hunk 19]
```diff
@@ -102,29 +102,26 @@ function it_should_remove_a_item_to_a_cart_from_event(CartItemEvent $event, Cart
...
         $event->getCart()->willReturn($cart);
         $manager->remove($cart)->shouldBeCalled();
         $manager->flush()->shouldBeCalled();
-        $provider->abandonCart()->shouldBeCalled();
 
         $this->clearCart($event);
     }
 
-    function it_should_save_a_valid_cart($manager, $provider, CartEvent $event, CartInterface $cart)
+    function it_should_save_a_valid_cart($manager, CartEvent $event, CartInterface $cart)
```
Comment (pamil) [inline src/Sylius/Bundle/CartBundle/spec/EventListener/CartSubscriberSpec.php]: If you're changing this line, replace `$manager` with `ObjectManager $manager` :)

File: src/Sylius/Bundle/CartBundle/EventListener/SessionCartSubscriber.php (hunk 17 above)
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/CartBundle/EventListener/SessionCartSubscriber.php]: Disable this subscriber on API requests :)

File: src/Sylius/Bundle/CartBundle/spec/EventListener/CartSubscriberSpec.php [hunk 20]
```diff
@@ -102,29 +102,26 @@ function it_should_remove_a_item_to_a_cart_from_event(CartItemEvent $event, Cart
         $this->removeItem($event);
     }
 
-    function it_should_clear_a_cart_from_event(CartEvent $event, CartInterface $cart, $manager, $provider)
+    function it_should_clear_a_cart_from_event(CartEvent $event, CartInterface $cart, $manager)
```
Comment (pamil) [inline src/Sylius/Bundle/CartBundle/spec/EventListener/CartSubscriberSpec.php]: If you're changing this line, replace `$manager` with `ObjectManager $manager` (and put it as the first argument) :)

File: src/Sylius/Bundle/CartBundle/Resources/config/services.xml (hunk 18 above)
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/CartBundle/Resources/config/services.xml]: In a separate PR, please 😉

File: src/Sylius/Bundle/CoreBundle/Cart/ItemResolver.php [hunk 21]
```diff
@@ -60,22 +60,22 @@ class ItemResolver implements ItemResolverInterface
     protected $channelContext;
 
     /**
-     * @param CartProviderInterface          $cartProvider
+     * @param CartContextInterface          $cartContext
      * @param RepositoryInterface            $productRepository
```
Comment (pamil) [inline src/Sylius/Bundle/CoreBundle/Cart/ItemResolver.php]: We may delete this whole indentation from this docblock.

File: src/Sylius/Bundle/CoreBundle/Cart/ItemResolver.php [hunk 22]
```diff
@@ -60,22 +60,22 @@ class ItemResolver implements ItemResolverInterface
...
      * @param FormFactoryInterface           $formFactory
      * @param AvailabilityCheckerInterface   $availabilityChecker
      * @param DelegatingCalculatorInterface  $priceCalculator
      * @param ChannelContextInterface        $channelContext
      */
     public function __construct(
-        CartProviderInterface          $cartProvider,
+        CartContextInterface          $cartContext,
         RepositoryInterface            $productRepository,
```
Comment (pamil) [inline src/Sylius/Bundle/CoreBundle/Cart/ItemResolver.php]: Same here, we can remove the alignment.

File: src/Sylius/Bundle/CoreBundle/Resources/config/doctrine/model/Channel.orm.xml (hunk 16 above)
Comment (pamil) [inline src/Sylius/Bundle/CoreBundle/Resources/config/doctrine/model/Channel.orm.xml]: Should be reverted.

**Comment by michalmarcinkowski** [comment]
Great job! Thanks @pjedrzejewski and @Arminek! 👍

**Comment by michalmarcinkowski** [comment]
@Arminek please apply all comments in a separate PR.

File: src/Sylius/Bundle/CoreBundle/spec/EventListener/UserCartRecalculationListenerSpec.php [hunk 23]
```diff
@@ -36,39 +40,40 @@ function it_is_initializable()
     }
 
     function it_recalculates_cart_for_logged_in_user(
-        CartProviderInterface $cartProvider,
+        CartContextInterface $cartContext,
         Event $event,
         OrderInterface $order,
         OrderRecalculatorInterface $orderRecalculator
```
Comment (pamil) [inline src/Sylius/Bundle/CoreBundle/spec/EventListener/UserCartRecalculationListenerSpec.php]: Should be the second one (it's the collaborator injected into constructor).

File: src/Sylius/Bundle/CoreBundle/spec/EventListener/UserCartRecalculationListenerSpec.php [hunk 24]
```diff
@@ -36,39 +40,40 @@ function it_is_initializable()
...
-        $this->recalculateCartWhileLogin($event);
-    }
-
     function it_throws_exception_if_provided_cart_is_not_order(
         CartInterface $cart,
-        CartProviderInterface $cartProvider,
-        Event $event
+        CartContextInterface $cartContext,
+        Event $event,
+        OrderRecalculatorInterface $orderRecalculator
```
Comment (pamil) [inline src/Sylius/Bundle/CoreBundle/spec/EventListener/UserCartRecalculationListenerSpec.php]: Should be the second one (it's the collaborator injected into constructor).

File: src/Sylius/Bundle/CoreBundle/spec/EventListener/UserCartRecalculationListenerSpec.php [hunk 25]
```diff
@@ -36,39 +40,40 @@ function it_is_initializable()
...
-        $cartProvider->getCart()->shouldNotBeCalled();
-
-        $this->recalculateCartWhileLogin($event);
-    }
-
     function it_throws_exception_if_provided_cart_is_not_order(
         CartInterface $cart,
-        CartProviderInterface $cartProvider,
-        Event $event
+        CartContextInterface $cartContext,
```
Comment (pamil) [inline src/Sylius/Bundle/CoreBundle/spec/EventListener/UserCartRecalculationListenerSpec.php]: Should be the first one.

File: src/Sylius/Bundle/WebBundle/Menu/FrontendMenuBuilder.php [hunk 26]
```diff
@@ -51,9 +51,9 @@ class FrontendMenuBuilder extends MenuBuilder
     /**
      * Cart provider.
```
Comment (pamil) [inline src/Sylius/Bundle/WebBundle/Menu/FrontendMenuBuilder.php]: Outdated and redundant comment.

File: src/Sylius/Component/Cart/Context/CartNotFoundException.php [hunk 27]
```diff
@@ -0,0 +1,26 @@
...
+ * @author Paweł Jędrzejewski <pawel@sylius.org>
+ */
+class CartNotFoundException extends \RuntimeException
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function __construct(\Exception $previousException = null)
+    {
+        parent::__construct('Sylius was not able to figure out the current cart.', 0, $previousException);
```
Comment (pamil) [inline src/Sylius/Component/Cart/Context/CartNotFoundException.php]: $message ?: 'Default message'

File: src/Sylius/Component/Cart/Context/CartNotFoundException.php [hunk 28]
```diff
@@ -0,0 +1,26 @@
...
+
+/**
+ * @author Paweł Jędrzejewski <pawel@sylius.org>
+ */
+class CartNotFoundException extends \RuntimeException
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function __construct(\Exception $previousException = null)
```
Comment (pamil) [inline src/Sylius/Component/Cart/Context/CartNotFoundException.php]: `__construct($message = null, \Exception $previousException = null)`

File: src/Sylius/Component/Cart/Context/CompositeCartContext.php [hunk 29]
```diff
@@ -0,0 +1,50 @@
...
+ */
+
+namespace Sylius\Component\Cart\Context;
+
+use Sylius\Component\Registry\PrioritizedServiceRegistryInterface;
+
+/**
+ * @author Arkadiusz Krakowiak <arkadiusz.krakowiak@lakion.com>
+ */
+class CompositeCartContext implements CartContextInterface
```
Comment (pamil) [inline src/Sylius/Component/Cart/Context/CompositeCartContext.php]: `final`?

File: src/Sylius/Component/Cart/composer.json [hunk 30]
```diff
@@ -23,6 +23,7 @@
         "php": "^5.6|^7.0",
 
         "sylius/order": "^1.0",
+        "sylius/registry": "^1.0",
```
Comment (pamil) [inline src/Sylius/Component/Cart/composer.json]: 😿

File: src/Sylius/Component/Cart/spec/Context/CartContextSpec.php [hunk 31]
```diff
@@ -0,0 +1,48 @@
...
+final class CartContextSpec extends ObjectBehavior
+{
+    function let(FactoryInterface $cartFactory)
+    {
+        $this->beConstructedWith($cartFactory);
+    }
+
+    function it_is_initializable()
+    {
+        $this->shouldHaveType('Sylius\Component\Cart\Context\CartContext');
```
Comment (pamil) [inline src/Sylius/Component/Cart/spec/Context/CartContextSpec.php]: We may use `::class`.

File: src/Sylius/Component/Cart/spec/Context/CompositeCartContextSpec.php (hunk 10 above)
Comment (pamil) [inline src/Sylius/Component/Cart/spec/Context/CompositeCartContextSpec.php]: Unfortunately unused :(

File: src/Sylius/Component/Cart/spec/Context/CompositeCartContextSpec.php [hunk 32]
```diff
@@ -0,0 +1,68 @@
...
+final class CompositeCartContextSpec extends ObjectBehavior
+{
+    function let(PrioritizedServiceRegistryInterface $prioritizedServiceRegistry)
+    {
+        $this->beConstructedWith($prioritizedServiceRegistry);
+    }
+
+    function it_is_initializable()
+    {
+        $this->shouldHaveType('Sylius\Component\Cart\Context\CompositeCartContext');
```
Comment (pamil) [inline src/Sylius/Component/Cart/spec/Context/CompositeCartContextSpec.php]: `::class`

File: src/Sylius/Component/Core/Cart/Context/ShopBasedCartContext.php [hunk 33]
```diff
@@ -0,0 +1,76 @@
...
+     */
+    private $cartFactory;
+
+    /**
+     * @var ShopperContextInterface
+     */
+    private $shopperContext;
+
+    /**
+     * @var CartInterface
```
Comment (pamil) [inline src/Sylius/Component/Core/Cart/Context/ShopBasedCartContext.php]: `CartInterface|null`

File: src/Sylius/Component/Core/Cart/Context/ShopBasedCartContext.php [hunk 34]
```diff
@@ -0,0 +1,76 @@
...
+        /** @var OrderInterface $cart */
+        $cart = $this->cartFactory->createNew();
+
+        try {
+            $cart->setChannel($this->shopperContext->getChannel());
+            $cart->setCurrencyCode($this->shopperContext->getCurrencyCode());
+        } catch (ChannelNotFoundException $exception) {
+            throw new CartNotFoundException($exception);
+        }
+        $cart->setCustomer($this->shopperContext->getCustomer());
```
Comment (pamil) [inline src/Sylius/Component/Core/Cart/Context/ShopBasedCartContext.php]: May we add a blank line above?

File: src/Sylius/Component/Core/Cart/Context/ShopBasedCartContext.php (hunk 6 above)
Comment (pamil) [inline src/Sylius/Component/Core/Cart/Context/ShopBasedCartContext.php]: Anyway, we may extract it as a decorator and use for different cart contexts too instead of always embedding this code.

File: src/Sylius/Component/Core/spec/Cart/Context/ShopBasedCartContextSpec.php [hunk 35]
```diff
@@ -0,0 +1,93 @@
...
+final class ShopBasedCartContextSpec extends ObjectBehavior
+{
+    function let(FactoryInterface $cartFactory, ShopperContextInterface $shopperContext)
+    {
+        $this->beConstructedWith($cartFactory, $shopperContext);
+    }
+
+    function it_is_initializable()
+    {
+        $this->shouldHaveType('Sylius\Component\Core\Cart\Context\ShopBasedCartContext');
```
Comment (pamil) [inline src/Sylius/Component/Core/spec/Cart/Context/ShopBasedCartContextSpec.php]: `::class`

File: src/Sylius/Bundle/CartBundle/Resources/config/services.xml (hunk 18 above)
Comment (Arminek) [inline src/Sylius/Bundle/CartBundle/Resources/config/services.xml]: We discuss about this change and i was going to do this in separate PR 😉

File: src/Sylius/Bundle/WebBundle/Menu/FrontendMenuBuilder.php (hunk 26 above)
Comment (Arminek) [inline src/Sylius/Bundle/WebBundle/Menu/FrontendMenuBuilder.php]: It is in webbundle i think i will left this one.

File: src/Sylius/Component/Cart/spec/Context/CartContextSpec.php (hunk 31 above)
Comment (Arminek) [inline src/Sylius/Component/Cart/spec/Context/CartContextSpec.php]: Did we change convention for initializable method?

Comment (pamil) [inline src/Sylius/Component/Cart/spec/Context/CartContextSpec.php]: After upgrade to PhpSpec 3.0 (#5541), where the default template uses `::class`.

File: src/Sylius/Bundle/CartBundle/Resources/config/services.xml (hunk 18 above)
Comment (Arminek) [inline src/Sylius/Bundle/CartBundle/Resources/config/services.xml]: Should i create for this separate service something like ``CartContextQueue``? or just simple add another method to the ``CompositeCartContext`` WDYT @pamil ?

Comment (pamil) [inline src/Sylius/Bundle/CartBundle/Resources/config/services.xml]: IMO you just need to copy-paste `CompositeChannelContext` and replace `channel` with `cart` :tada: There's nothing to customize :)
